numpy/core/src/private/binop_override.h

```diff
@@ -0,0 +1,196 @@
+     * logic for forward binop implementations.
+     */
+
+    /*
+     * NB: there's another copy of this code in
+     *    numpy.ma.core.MaskedArray._delegate_binop
+     * which should possibly be updated when this is.
+     */
+
+    PyObject *attr;
+    double self_prio, other_prio;
+    int defer;
+    /*
+     * attribute check is expensive for scalar operations, avoid if possible
+     */
+    if (other == NULL ||
+        self == NULL ||
+        Py_TYPE(self) == Py_TYPE(other) ||
+        PyArray_CheckExact(other) ||
+        PyArray_CheckAnyScalarExact(other) ||
+	_is_basic_python_type(other)) {
+        return 0;
+    }
+    /*
+     * Classes with __array_ufunc__ are living in the future, and only need to
+     * check whether __array_ufunc__ equals None.
+     */
+    attr = PyArray_GetAttrString_SuppressException(other, "__array_ufunc__");
+    if (attr) {
+        defer = (attr == Py_None);
+        Py_DECREF(attr);
+        return defer;
+    }
+    /*
+     * Otherwise, we need to check for the legacy __array_priority__. But if
+     * other.__class__ is a subtype of self.__class__, then it's already had
+     * a chance to run, so no need to defer to it.
+     */
+    if(PyType_IsSubtype(Py_TYPE(other), Py_TYPE(self))) {
+        return 0;
+    }
+    self_prio = PyArray_GetPriority((PyObject *)self, NPY_SCALAR_PRIORITY);
+    other_prio = PyArray_GetPriority((PyObject *)other, NPY_SCALAR_PRIORITY);
+    return self_prio < other_prio;
+}
+
+/*
+ * A CPython slot like ->tp_as_number->nb_add gets called for *both* forward
+ * and reversed operations. E.g.
+ *   a + b
+ * may call
+ *   a->tp_as_number->nb_add(a, b)
+ * and
+ *   b + a
+ * may call
+ *   a->tp_as_number->nb_add(b, a)
+ * and the only way to tell which is which is for a slot implementation 'f' to
+ * check
+ *   arg1->tp_as_number->nb_add == f
+ *   arg2->tp_as_number->nb_add == f
+ * If both are true, then CPython will as a special case only call the
+ * operation once (i.e., it performs both the forward and reversed binops
+ * simultaneously). This function is mostly intended for figuring out
+ * whether we are a forward binop that might want to return NotImplemented,
+ * and in the both-at-once case we never want to return NotImplemented, so in
+ * that case BINOP_IS_FORWARD returns false.
+ *
+ * This is modeled on the checks in CPython's typeobject.c SLOT1BINFULL
+ * macro.
+ */
+#define BINOP_IS_FORWARD(m1, m2, SLOT_NAME, test_func)  \
+    (Py_TYPE(m2)->tp_as_number != NULL &&                               \
+     (void*)(Py_TYPE(m2)->tp_as_number->SLOT_NAME) != (void*)(test_func))
+
+#define BINOP_GIVE_UP_IF_NEEDED(m1, m2, slot_expr, test_func)           \
+    do {                                                                \
+        if (BINOP_IS_FORWARD(m1, m2, slot_expr, test_func) &&           \
+            binop_override_forward_binop_should_defer((PyObject*)m1, (PyObject*)m2)) { \
```

Comment:
I think you're right that with `__array_ufunc__ = None`, the reverse binop might as well return `NotImplemented` immediately, although in practice the result is the same; see the comments at https://github.com/numpy/numpy/pull/8247/files#diff-a76a87b213692b2a554a63aebea985caR65

It does make me whether we're not better off calling into `ndarray.__array_ufunc__` -- we seem to be duplicating logic in multiple places.